File: app/src/main/java/com/luffy/mulmedia/activity/NavigationActivity.java
package com.luffy.mulmedia.activity;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Rect;
import android.net.Uri;
import android.os.Bundle;
import android.os.ParcelFileDescriptor;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.luffy.mulmedia.R;
import com.luffyxu.mulmedia.model.NavItem;
import com.luffyxu.mulmedia.ui.adapter.NavItemAdapter;

import java.util.ArrayList;
import java.util.List;

import kotlin.Unit;
import kotlin.jvm.functions.Function2;

public class NavigationActivity extends AppCompatActivity {

    private static final String TAG = "NavigationActivity";
    RecyclerView mLessonList;
    List<NavItem> mLessonPaths;

    NavItemAdapter adapter;

    Button mSelectFileBtn;
    EditText mSelectFileEt;

    private Uri selectUri;
    private ParcelFileDescriptor descriptor;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_navigation);
        mLessonList = findViewById(R.id.LessonListView);
        mSelectFileBtn = findViewById(R.id.btn_select_file);
        mSelectFileEt = findViewById(R.id.et_select_file_uri);
        mSelectFileBtn.setOnClickListener((view) -> {
            if(checkPermission()) {
                startFileBrowser();
            }else {
                requestPermission();
            }
        });
    }

    private boolean checkPermission(){
        int code = ActivityCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE);
        if(code == PackageManager.PERMISSION_GRANTED){
            return true;
        }else {
            return false;
        }
    }

    private void requestPermission(){
        ActivityCompat.requestPermissions(this,new String[]{Manifest.permission.READ_EXTERNAL_STORAGE},1);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if(requestCode == 0){
            if(grantResults[0] == PackageManager.PERMISSION_GRANTED){
                Log.d(TAG,"权限申请成功");
            }else {
                Log.d(TAG,"权限申请失败");
            }
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        mLessonPaths = createNavList();
        adapter = new NavItemAdapter();
        adapter.setData(mLessonPaths);
        adapter.setItemClickListener(new Function2<View, Integer, Unit>() {
            @Override
            public Unit invoke(View view, Integer integer) {
                navigate(integer);
                return null;
            }
        });
        mLessonList.setAdapter(adapter);
        mLessonList.addItemDecoration(new RecyclerView.ItemDecoration() {
            @Override
            public void getItemOffsets(@NonNull Rect outRect, int itemPosition, @NonNull RecyclerView parent) {
                super.getItemOffsets(outRect, itemPosition, parent);
                outRect.left = 20;
                outRect.right = 20;
                outRect.bottom = 20;
            }
        });

        mLessonList.setLayoutManager(new LinearLayoutManager(this,RecyclerView.VERTICAL,false));
    }

    private List<NavItem> createNavList(){
        String[] pathArrays = getResources().getStringArray(R.array.media_lesson_path_collection);
        String[] titleArrays = getResources().getStringArray(R.array.media_lesson_title_collection);
        List<NavItem> data = new ArrayList<>();
        for(int i=0;i< pathArrays.length;i++){
            data.add(new NavItem(titleArrays[i],pathArrays[i]));
        }
        return data;
    }

    private void navigate(int position){
//        if(selectUri == null){
//            Toast.makeText(this, "please select a video first", Toast.LENGTH_SHORT).show();
//            return;
//        }
        String path = mLessonPaths.get(position).getPath();
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse("xzq://navigate"+path));
        intent.putExtra("uri",selectUri);
        startActivity(intent);
    }

    private void startFileBrowser(){
        Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
        intent.setType("video/*");
        startActivityForResult(intent,1);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if(requestCode == 1){
            Uri uri = data.getData();
            Log.d(TAG,"onActivityResult " + uri);
            selectUri = uri;
            mSelectFileEt.setText(uri.toString());
        }
    }
}

File: app/src/main/java/com/luffy/mulmedia/activity/NavigationActivity.kt
package com.luffy.mulmedia.activity

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Rect
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import androidx.core.app.ActivityCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.luffy.mulmedia.R
import com.luffy.mulmedia.consts.Consts.titleArray
import com.luffy.mulmedia.databinding.ActivityNavigationBinding
import com.luffyxu.mulmedia.model.NavItem
import com.luffyxu.mulmedia.ui.adapter.NavItemAdapter
import java.io.FileDescriptor

class NavigationActivity : BaseActivity() {
    var mLessonPaths: MutableList<NavItem> = mutableListOf()
    lateinit var adapter: NavItemAdapter
    private var selectUri: Uri? = null
    lateinit var binding: ActivityNavigationBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityNavigationBinding.inflate(LayoutInflater.from(this))
        setContentView(binding.root)

        mLessonPaths = createNavList()
        adapter = NavItemAdapter().apply {
            data.addAll(mLessonPaths)
            itemClickListener = { view: View?, integer: Int ->
                if (integer == 0) {
                    if (checkPermission()) {
                        startFileBrowser()
                    } else {
                        requestPermission()
                    }
                } else {
                    navigate(integer)
                }
            }
        }
        binding.LessonListView.apply {
            adapter = this@NavigationActivity.adapter
            addItemDecoration(object : RecyclerView.ItemDecoration() {
                override fun getItemOffsets(
                    outRect: Rect,
                    itemPosition: Int,
                    parent: RecyclerView
                ) {
                    super.getItemOffsets(outRect, itemPosition, parent)
                    outRect.left = 20
                    outRect.right = 20
                    outRect.bottom = 20
                }
            })
            layoutManager =
                LinearLayoutManager(this@NavigationActivity, RecyclerView.VERTICAL, false)
        }
    }

    private fun checkPermission(): Boolean {
        val code =
            ActivityCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE)
        return code == PackageManager.PERMISSION_GRANTED
    }

    private fun requestPermission() {
        ActivityCompat.requestPermissions(
            this,
            arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE),
            1
        )
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == 0) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                Log.d(TAG, "权限申请成功")
            } else {
                Log.d(TAG, "权限申请失败")
            }
        }
    }

    override fun onResume() {
        super.onResume()
    }

    private fun createNavList(): MutableList<NavItem> {
        val pathArrays = resources.getStringArray(R.array.media_lesson_path_collection)
        val titleArrays = titleArray()
        val data: MutableList<NavItem> = ArrayList()
        for (i in titleArrays.indices) {
            if (i == 0) {
                data.add(NavItem(titleArrays[i], ""))
            } else {
                data.add(NavItem(titleArrays[i], pathArrays[i - 1]))
            }
        }
        return data
    }

    private fun navigate(position: Int) {
        val path = mLessonPaths[position].path
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse("xzq://navigate$path"))
        intent.putExtra("uri", selectUri)
        startActivity(intent)
    }

    private fun startFileBrowser() {
        val intent = Intent(Intent.ACTION_OPEN_DOCUMENT)
        intent.type = "video/*"
        startActivityForResult(intent, 1)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == 1) {
            val uri = data!!.data
            Log.d(TAG, "onActivityResult $uri")
            selectUri = uri
            adapter.data[0].description = uri.toString()
            adapter.notifyItemChanged(0)
        }
    }

    override fun onUriAction(uri: Uri?) {
    }

    override fun onUriAction(uri: FileDescriptor?) {
    }

    companion object {
        private const val TAG = "NavigationActivity"
    }
}